src/devices/src/virtio/net/device.rs

```diff
@@ -141,14 +141,19 @@ impl Net {
             | 1 << VIRTIO_F_VERSION_1;
 
         let mut config_space = ConfigSpace::default();
-        if let Some(mac) = guest_mac {
-            config_space.guest_mac.copy_from_slice(mac.get_bytes());
+        let mac: MacAddr;
+        if let Some(guest_mac) = guest_mac {
+            config_space
+                .guest_mac
+                .copy_from_slice(guest_mac.get_bytes());
             // When this feature isn't available, the driver generates a random MAC address.
             // Otherwise, it should attempt to read the device MAC address from the config space.
             avail_features |= 1 << VIRTIO_NET_F_MAC;
-        }
 
-        let guest_mac = guest_mac.copied();
+            mac = *guest_mac;
+        } else {
+            mac = MacAddr::from_bytes_unchecked(&config_space.guest_mac[..MAC_ADDR_LEN]);
```

Comment:
My understanding is that if the config space is not given, it will be set at device activation time. This means that when we configure/build a network device without the `guest_mac` information, this information remains unknown until the network device gets activated. I think we still need `Option<MacAddr>` for guest_mac field of the Net device.

To have the `guest_mac` cache in sync with the content from the config space, we would need to update this cache when this content is change. This is done in `write_config` implementation of `VirtioDevice` trait for `Net`.

Reply:
I was hoping that we can get rid of the Option, but you're right, we have to know if we have a valid MAC address. 
For the save/restore case, should we save the config space *and* the Option<MacAddr> or can we assume that we save the device after the microvm booted and we have a valid address?

Reply:
For the save/restore case, as we've discussed offline, it is sufficient to save only the config space and populate at restore time the `guest_mac` as well, besides the config space.

It is important to note that this parsing of the config space, to update the guest_mac cache, is done in the device activation, as well in the device restore, which are on the hotpath as well. The tradeoff is that we will have no degradation in the network device performance, but we will have some degradation on the microVM booting/restoring from snapshot. I do not know what to expect from this degradation, and if the performance tests are green, I guess it is not significant at all.